pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,425 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+	// each time we add a List or Counter. Start at point where desired replicas was determined based
+	// on case where len(list.Values) or counter.Count == 0.
+	aggCount += count * additionalReplicas
+	aggCapacity += capacity * additionalReplicas
+	limited := false
+	for {
+		// Check if we've reached MaxCapacity
+		if aggCapacity == maxCapacity {
+			limited = true
+			break
+		}
+		if aggCapacity > maxCapacity {
+			limited = true
+			replicas--
+			break
+		}
+		// Check if we've reached desiredCapacity
+		desiredCapacity = (float64(aggCount) * 100) / (100 - bufferPercent)
+		desiredReplicas := math.Ceil(desiredCapacity / float64(capacity))
+		if replicas >= int32(desiredReplicas) {
+			break
+		}
+		// Keep checking if adding one List or Counter will reach our desiredCapacity
+		aggCount += count
+		aggCapacity += capacity
+		replicas++
+	}
+
+	return replicas, limited, nil
+}
+
+// scaleDownByPercent
+// TODO: This death-spirals down to min capacity, so the customer would need to have something
+// in place to prevent gameservers from being evicted elsewhere if they have Count / Values on
+// them. I'm assuming this is the behavior we intend, and it just needs good documentation?
+func scaleDownByPercent(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount, key string, isCounter bool, replicas int32,
+	aggCount, aggCapacity, minCapacity int64, bufferPercent float64) (int32, bool, error) {
+	// Exit early if we're already at MinCapacity to avoid calling getSortedGameServers if unnecessary
+	if aggCapacity == minCapacity {
+		return replicas, true, nil
+	}
+
+	// We first need to get the individual game servers in order of deletion on scale down, as both
+	// the Count and Capacity can change, so we do not know from the aggregate counts how much
+	// removing x game servers will affect the aggregate count and capacity.
+	gameServers, err := getSortedGameServers(f, gameServerLister, nodeCounts)
+	if err != nil {
+		return 0, false, err
+	}
+
+	// Determine the desiredCapacity based on removing one gameserver
+	limited := false
+	var desiredCapacity int64
+	for _, gs := range gameServers {
+		// Check if we've reached desiredCapacity
+		desiredCapacity = int64(math.Ceil((float64(aggCount) * 100) / (100 - bufferPercent)))
+		if desiredCapacity >= aggCapacity {
+			break
+		}
+		// Keep checking if adding removing one Counter or List will reach our desiredCapacity
+		switch isCounter {
+		case true:
+			if counter, ok := gs.Status.Counters[key]; ok {
+				aggCount -= counter.Count
+				aggCapacity -= counter.Capacity
+			}
+		case false:
+			if list, ok := gs.Status.Lists[key]; ok {
+				aggCount -= int64(len(list.Values))
+				aggCapacity -= list.Capacity
+			}
+		}
+		replicas--
+		// Check if we've reached MinCapacity
+		if aggCapacity == minCapacity {
+			limited = true // TODO: Should we have this return true or false when scaling down to 0 (MinCapacity == 0)?
```

Comment:
Should we have this return true or false when scaling down to 0 (MinCapacity == 0)?

Reply:
If I use what currently exists as a reference:

https://github.com/googleforgames/agones/blob/cef40425076e5e9a766952eb2f499d9b447b703a/pkg/fleetautoscalers/fleetautoscalers.go#L216-L223

I think that means it should return false to be consistent. Do you read it the same? (extra eyes good).

Reply:
It's not quite the same -- the bufferPolicy is able to calculate the desiredReplicas and then determine if it's limited or not. 

We're checking after removing each unique gameserver one by one if we've reached the limit. If we've reached our desired state on line 624 then we exit before checking if it's limited. So in our case we would almost always hit aggCapacity == minCapacity if we have not reached the desired capacity, yet we've hit our minimum. I think the only case where this line might return "false" is when remove every single game server (so we don't go back up into the loop to check desiredCapacity again), but we've hit minCapacity because we've removed all the gameservers.

Alternatively I could add an emtpy gameserver at the end of the gameserver list to get it to loop one last time (after all gameservers have been removed). Would that be a reasonable thing to do?

Reply:
I would apply the same logic here, as we do elsewhere: if the _desired_ capacity is less than the current aggregate allocated capacity - we are limited. 

So I think we should be looking at `desiredCapacity` here, not aggCapacity (Aggregate capacity?)